perllib/FixMyStreet/App/Controller/Report.pm

```diff
@@ -660,6 +662,59 @@ sub map :Chained('id') :Args(0) {
     $c->res->body($image->{data});
 }
 
+sub confirmation : Path('confirmation') : Args(1) {
+    my ( $self, $c, $id ) = @_;
+
+    # First of all check that the report ID is valid
+    my $report = FixMyStreet::DB->resultset('Problem')->find( { id => $id } );
+    unless ( $report ) {
+        $c->detach( '/page_error_404_not_found', [] );
+    }
+
+    # Now verify the token we've been given is correct
+    my $token_param = $c->get_param('token') || "";
+    my $token = $c->forward('/report/get_confirmation_token', [ $report ]);
+    unless ( $token eq $token_param ) {
+        $c->detach( '/page_error_404_not_found', [] );
+    }
+
+    # If the token is valid but expired then may as well be helpful and bounce
+    # the user to report page rather than 404.
+    # (NB the report may still be unconfirmed, but end result is the same - a 404)
+    my $cutoff = DateTime->now()->subtract( minutes => 1 )->epoch;
+    my $timestamp = $report->confirmed ? $report->confirmed : $report->created;
+    if ( $timestamp->epoch < $cutoff ) {
+        return $c->res->redirect($report->url);
+    }
+
+    # We're now confident the user is allowed to view this report so stick it on
+    # the stash and load up the correct template.
+    $c->stash->{problem} = $report;
+    if ( $report->confirmed ) {
+        $c->stash->{template} = 'tokens/confirm_problem.html';
+        $c->stash->{created_report} = "loggedin";
+        $c->stash->{report} = $c->stash->{problem};
+    } else {
+        $c->stash->{template} = 'email_sent.html';
+        $c->stash->{email_type} = 'problem';
+    }
+}
+
+sub get_confirmation_token : Private {
```

Comment:
Just wondered if this would be better as a method on Problem.pm - you don't need $c as you can call `FixMyStreet::DB->resultset('Secret')->get`.